src/nvme/linux.c

```diff
@@ -543,6 +559,187 @@ int nvme_gen_dhchap_key(char *hostnqn, enum nvme_hmac_alg hmac,
 	HMAC_CTX_free(hmac_ctx);
 	return err;
 }
+
+static unsigned char *derive_retained_key(const EVP_MD *md, const char *hostnqn,
+					  unsigned char *generated_key,
+					  size_t key_len)
+{
+	unsigned char *retained_key;
+	EVP_PKEY_CTX *ctx;
+	size_t retained_len;
+	int err;
+
+	retained_key = malloc(key_len);
+	if (!retained_key)
+		return NULL;
+
+	retained_len = key_len;
+	ctx = EVP_PKEY_CTX_new_id(EVP_PKEY_HKDF, NULL);
+	if (!ctx)
+		goto out_free_retained_key;
+
+	err = -ENOKEY;
+	if (EVP_PKEY_derive_init(ctx) <= 0)
+		goto out_free_retained_key;
+	if (EVP_PKEY_CTX_set_hkdf_md(ctx, md) <= 0)
+		goto out_free_retained_key;
+	if (EVP_PKEY_CTX_set1_hkdf_key(ctx, generated_key, key_len) <= 0)
+		goto out_free_retained_key;
+	if (EVP_PKEY_CTX_add1_hkdf_info(ctx, "tls13 ", 6) <= 0)
+		goto out_free_retained_key;
+	if (EVP_PKEY_CTX_add1_hkdf_info(ctx, "HostNQN", 7) <= 0)
+		goto out_free_retained_key;
+	if (EVP_PKEY_CTX_add1_hkdf_info(ctx, hostnqn, strlen(hostnqn)) <= 0)
+		goto out_free_retained_key;
+
+	if (EVP_PKEY_derive(ctx, retained_key, &retained_len) <= 0) {
+		fprintf(stderr, "EVP_KDF derive failed\n");
```

Comment:
Printing to `stderr` for a library is not ideal. We have `nvme_msg` but without a context it would also print to stderr.  That's why the other OpenSSL code doesn't do any printing at all.

Reply:
Will be replacing it with nvme_msg() or removing it, as appropriate.